content/shmem_init_thread.tex

```diff
@@ -0,0 +1,78 @@
+\apisummary{
+Initializes the \openshmem{} library, similar to SHMEM\_INIT(), and in addition
+perform the initialization required for thread-safe invocation of \openshmem{} functions.}
+
+\begin{apidefinition}
+
+\begin{Csynopsis}
+int shmem_init_thread(int requested, int *provided);
+\end{Csynopsis}
+
+\begin{Fsynopsis}  
+INTEGER requested, provided
+     
+SHMEM_INIT_THREAD(requested, provided)
+\end{Fsynopsis}  
+
+\begin{apiarguments}
+\apiargument{IN}{requested}  {The thread level support requested by the user.
+The correct values are SHMEM\_THREAD\_SINGLE, SHMEM\_THREAD\_FUNNELED, or SHMEM\_THREAD\_MULTIPLE}
+\apiargument{OUT}{provided}{The thread level support provided by the \openshmem{} implementation.}
+\end{apiarguments}
+
+
+\apidescription{
+    \FUNC{shmem\_init\_thread}  initializes the \openshmem{} library similar to
+    \FUNC{shmem\_init}, and in addition perform the initialization required for
+    thread-safe invocation of \openshmem{} functions. The argument
+    \textit{requested} is used to specify the desired level of thread support.
+    The function returns the support provided by the library. There are four 
+    levels of threading support.
+ 
+ \begin{itemize}
+\item SHMEM\_THREAD\_SINGLE: The \openshmem{} program may not be multithreaded, 
+and all \openshmem{} interfaces are invoked by a single thread. 
+
+\item SHMEM\_THREAD\_FUNNELED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces.
+
+\item SHMEM\_THREAD\_SERIALIZED: 
+The \openshmem{} program may be multithreaded, however, the 
+program must ensure that only one thread invokes the \openshmem{}
+interfaces at any given instance i.e., the \openshmem{} interfaces 
+are not invoked concurrently by multiple threads.
+
+\item SHMEM\_THREAD\_MULTIPLE: 
+The \openshmem{} program may be multithreaded, and any 
+thread may invoke the \openshmem{} interfaces.
+
+\item SHMEM\_THREAD\_UNSUPPORTED: The \openshmem{} specification 
```

Comment:
Summary of discussion and options (please correct me , if I missed any options @nspark @spotluri @jdinan )

We need to define that there is a default mode when we initialize with shmem_init, and then we have to define the default mode. The three options for the default mode are:  

1. The default mode has to be one of the four modes (SINGLE,FUNNELED,SERIALIZED,MULTIPLE). 
2. The default mode has to be one of the four modes or a “no thread support mode” where only a subset of shmem interfaces are thread safe. 
3. The default mode is the extra mode - (UNSUPPORTED/UNDEFINED/IMPLEMENTATION_DEFINED)